internal/admission/validator_test.go

```diff
@@ -288,4 +292,185 @@ func TestKongHTTPValidator_ValidateClusterPlugin(t *testing.T) {
+				}(),
+			},
+			operation:       admissionv1.Create,
+			expectError:     true,
+			expectedMessage: "consumer credential failed validation",
+		},
+		{
+			name: "consumer refers a secret that is already referred by another consumer",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets: []*corev1.Secret{validSecret()},
+			consumers: []*configurationv1.KongConsumer{
+				func() *configurationv1.KongConsumer {
+					c := basicConsumer()
+					c.Name = "consumer-2"
+					c.Credentials = []string{"secret"}
+					return &c
+				}(),
+			},
+			operation:       admissionv1.Create,
+			expectError:     true,
+			expectedMessage: ErrTextConsumerCredentialUniqueKeyConstraintFailed,
+		},
+		{
+			name: "consumer refers a secret that has the same key as another secret",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets: []*corev1.Secret{
+				validSecret(),
+				func() *corev1.Secret {
+					s := validSecret()
+					s.Name = "secret-2"
+					return s
+				}(),
+			},
+			consumers: []*configurationv1.KongConsumer{
+				func() *configurationv1.KongConsumer {
+					c := basicConsumer()
+					c.Name = "consumer-2"
+					c.Credentials = []string{"secret-2"}
+					return &c
+				}(),
+			},
+			operation:       admissionv1.Create,
+			expectError:     true,
+			expectedMessage: ErrTextConsumerCredentialUniqueKeyConstraintFailed,
+		},
+		{
+			name: "consumer is updated in place - no unique constraint violation should occur",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets: []*corev1.Secret{
+				validSecret(),
+			},
+			consumers: []*configurationv1.KongConsumer{
+				func() *configurationv1.KongConsumer {
+					c := basicConsumer()
+					c.Credentials = []string{"secret"}
+					return &c
+				}(),
+			},
+			operation: admissionv1.Update,
+			expectOK:  true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+
+		t.Run(tt.name, func(t *testing.T) {
+			s, _ := store.NewFakeStore(store.FakeObjects{
+				Secrets:       tt.secrets,
+				KongConsumers: tt.consumers,
+			})
+			validator := KongHTTPValidator{
+				// For the sake of tests we use the same store for both SecretGetter and Store as it does not matter here
+				// if secrets are managed by us or not yet.
+				Store:               s,
+				SecretGetter:        s,
+				ConsumerSvc:         &fakeConsumerService{consumerAlreadyExists: tt.consumerAlreadyExistsInGateway},
+				ingressClassMatcher: fakeClassMatcher,
+			}
+
+			toValidate := basicConsumer()
+			tt.modifyBasicConsumer(&toValidate)
+			ok, message, err := validator.ValidateConsumer(context.Background(), toValidate, tt.operation)
+			assert.Equal(t, tt.expectOK, ok)
+			if tt.expectError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+			assert.Equal(t, tt.expectedMessage, message)
```

Comment:
This touches implementation but I thought I'd mention it here as it's visible how it surfaces here: I haven't followed the call stack on how the `message` is being used that's returned from `ValidateConsumer()` but I'd prefer us to use the error message 
for this purpose: to convey what is wrong.

Just a loose proposal in case the message does not have a specific use case up the stack somewhere.

Reply:
It's good you've noticed that. I think we had incorrect usages of the `(ok, msg, err)` trio in the validator. Up the stack, the error is used to respond to the API server with `code=500`, ignoring the message completely. If we have a regular validation error, we should not return an error, but just the message and `ok=false` as it will result in a `code=400` and the message being passed to the API server. 

I reviewed the validator code for that, also enhancing the messages to include detailed reasons when possible: [7adda5e](https://github.com/Kong/kubernetes-ingress-controller/pull/3239/commits/7adda5e4ed450f16e63a5c21eb3f6f69707a67f3)